# 01. Importing libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
path = r'C:\Users\s.wienecke\Documents\Fortbildung\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\s.wienecke\\Documents\\Fortbildung\\Instacart Basket Analysis'

In [4]:
# df stands for order.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [5]:
# df_prods stands for products.csv 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 03. Dropping columns

In [6]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

# 04. Identifying missing values

In [7]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

# 05. Rename columns

In [8]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [9]:
 df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 06. Changing a Variable’s Data Type

In [10]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

# 07. Importing data

In [11]:
# df_dep stands for departments.csv 
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [12]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

# 08. Transposing data

In [13]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [14]:
#  df_dep_t = creating a new dataframe from the transposed version
df_dep_t = df_dep.T

In [15]:
#  add an index to a dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [16]:
# create a new header
new_header = df_dep_t.iloc[0]

In [17]:
new_header

0    department
Name: department_id, dtype: object

In [18]:
#Remove the first row in the dataframe
df_dep_t_new = df_dep_t[1:]

In [19]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [20]:
# add new header
df_dep_t_new.columns = new_header

# 09. Data Dictionaries

In [21]:
data_dict = df_dep_t_new.to_dict('index')

In [22]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [23]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# 10. Subsetting

In [24]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [25]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [26]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [27]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [28]:
# second way of achieving these same results
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [29]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]


In [30]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [31]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# 4.4 Exercise

#  2. Task

In [32]:
# descriptive statistics for finding an identifier variable
df_ords.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [33]:
# changing numeric variable into string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [34]:
#check format
df_ords['user_id'].dtypes

dtype('O')

# 3. Task

In [35]:
# Renaming column
df_ords.rename(columns = {'order_hour_of_day' : 'order_time'}, inplace = True)

In [36]:
df_ords.head()

order_id user_id eval_set  order_number  orders_day_of_week  order_time  \
0  2539329       1    prior             1                   2           8   
1  2398795       1    prior             2                   3           7   
2   473747       1    prior             3                   3          12   
3  2254736       1    prior             4                   4           7   
4   431534       1    prior             5                   4          15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 4. Task

In [37]:
# max order nummer
df_ords['order_number'].max()

100

In [38]:
# new df with only max order number
df_ords_bussiest_hour = df_ords.loc[df_ords['order_number'] > 99]

In [39]:
# mean order time of max order number
df_ords_bussiest_hour['order_time'].mean()

12.950509461426492

# 5. Task

In [40]:
# meaning behind a value of 4 in the "department_id"
print(data_dict.get('4'))

{'department': 'produce'}


# 6. Task

In [41]:
#subset of breakfast item sales (14)
df_prods[df_prods['department_id']==14]

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# 7. Task

In [42]:
# subset of dinner parties items (alcohol, deli, beverages, and meat/seafood)
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [43]:
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# 8. Task

In [44]:
# number of rows
df_dinner_parties.shape

(7650, 5)

7650 rows has the dataframe df_dinner_parties.

# 9. Task

In [45]:
# select informtion about user_id 1
df_ords_user_id = df_ords.loc[df_ords['user_id'] == "1"]

In [46]:
df_ords_user_id

order_id user_id eval_set  order_number  orders_day_of_week  order_time  \
0   2539329       1    prior             1                   2           8   
1   2398795       1    prior             2                   3           7   
2    473747       1    prior             3                   3          12   
3   2254736       1    prior             4                   4           7   
4    431534       1    prior             5                   4          15   
5   3367565       1    prior             6                   2           7   
6    550135       1    prior             7                   1           9   
7   3108588       1    prior             8                   1          14   
8   2295261       1    prior             9                   1          16   
9   2550362       1    prior            10                   4           8   
10  1187899       1    train            11                   4           8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# 10. Task

In [47]:
# basic stats: max order time
df_ords_user_id['order_time'].max()

16

In [48]:
# basic stats: mean orders day of week
df_ords_user_id['orders_day_of_week'].mean()

2.6363636363636362

In [49]:
# basic stats: min days since prior order
df_ords_user_id['days_since_prior_order'].min()

0.0

# 11. Task

In [50]:
# export df_ords to folder Prepared Data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [ ]:
# export df_dep_t_new to folder Prepared Data
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))